python/predict_urls/prediction_funcs.py

```diff
@@ -0,0 +1,280 @@
+            #     deleted_df = pd.concat([deleted_df, pd.DataFrame([deleted_row])], ignore_index=True)
+            #     df = df.drop(idx)
+            #     df.to_csv(OPD_SOURCE_TABLE, index=False)
+            #     deleted_df.to_csv(DELETED_TABLE, index=False)
+            #     continue
+        else:
+            valid, new_url = find_valid_url_for_year(url, y, year_str, data_type, spreadsheet_fields)
+            if valid:
+                if verbose:
+                    print(f"Valid URL found: {new_url}. Adding to OPD_Source_table.")
+                # Prepare row for OPD_Source_table
+                fields = [
+                    "State", "SourceName", "Agency", "AgencyFull", "TableType",
+                    "coverage_start", "coverage_end", "last_coverage_check", "Description",
+                    "source_url", "readme", "URL", "Year", "DataType", "date_field",
+                    "dataset_id", "agency_field", "min_version", "query"
+                ]
+                row = {f: "" for f in fields}
+                row["URL"] = new_url
+                row["Year"] = str(y)
+                row["last_coverage_check"] = datetime.now().strftime("%m/%d/%Y")
+                row["coverage_start"] = f"01/01/{y}"
+                if y == current_year:
+                    row["coverage_end"] = datetime.now().strftime("%m/%d/%Y")
+                else:
+                    row["coverage_end"] = f"12/31/{y}"
+                if spreadsheet_fields:
+                    for k, v in spreadsheet_fields.items():
+                        if k in row:
+                            row[k] = v
+                # Append to OPD_Source_table.csv
+                df = pd.read_csv(OPD_SOURCE_TABLE)
+                df = pd.concat([df, pd.DataFrame([row])], ignore_index=True)
+                df.to_csv(OPD_SOURCE_TABLE, index=False)
+                if verbose:
+                    print(f"Added to OPD_Source_table: {new_url}")
+                continue
+            else:
+                if verbose:
+                    print(f"{new_url}: not valid. Skipping.")
+    return None
+
+def auto_update_sources(
+    outdated_days=None,
+    verbose=False
+):
+    """
+    Automatically check for new data by incrementing year in URLs for testable sources. Adds valid URLs to OPD_Source_table.csv.
+
+    Args:
+        outdated_days (int): How many days before a source is considered outdated.
+        verbose (bool): Print progress.
+    """
+    # 1. Load table and parse last_coverage_check as datetime
+    df = pd.read_csv(OPD_SOURCE_TABLE)
+    df["last_coverage_check_dt"] = pd.to_datetime(df["last_coverage_check"], errors="coerce")
+    # 2. Group by composite key and get max last checked date
+    composite_key = ['State', 'SourceName', 'AgencyFull', 'TableType']
+    max_df = df.loc[df.groupby(composite_key)["Year"].idxmax()]
+    # 3. Filter URLs with exactly one 4-digit year
+    max_df = max_df[max_df["URL"].str.contains(r"\d{4}", regex=True)]
```

Comment:
We could probably be more constraining in our search here. We actually know the specific year that we are searching for because it's in the Year column. We could search for that rather than 4 consecutive numbers. If we do this, I don't think there will be any need for the below line (236)